posts/2018-12-11-hunt-bugs-down-before-they-are-merged.md

```diff
@@ -0,0 +1,362 @@
+
+As the VP leaves, you sing to yourself:
+
+> ... there's a moooooon over bourbon street ...
+
+</div>
+
+Most of the time no one knows where the bug is and who introduced it. This
+means that the odds of picking the right engineer to hunt down a bug are about
+the same as trying to guess who in your team got their tonsils removed last --
+if anybody actually got their tonsils removed.
+
+The person tasked with fixing the bug then is not particularly likely to know
+about the part of the codebase that is at fault. Even if the person who
+introduced the bug is working on hunting it and fixing it, that person might
+not have worked on that part of the codebase for a while. The context is either
+absent or has been lost with time, which make debugging longer and more
+frustrating.
+
+This can be avoided by catching the bugs right after they've been written, and
+before they've been merged.
+
+### Moar noise
+
+<div class="story">
+
+> you> @channel I've got a fix for the server crash issue, anybody care to
+> review #2578?
+>
+> tom> please use @here instead of @channel
+>
+> pat> @tom please don't use "at" channel
+>
+> jen> @you yes I'll have a look in a sec
+>
+> mol> @you sure np
+>
+> mkt> @channel coffee break?
+
+</div>
+
+<div style="display: inline-block;">
+<figure style="float: left;">
+<img src="../images/where_does_it_end.jpg" alt="Where does it end"><figcaption>Where does it end?</figcaption>
+</figure>
+Fixed the issue? A new round of PRs and reviews creates even more noise. At
+best this justifies hiring the intern whose job it is to move JIRA tickets
+around. Most likely it's a waste of time for everyone involved.
+</div>
+
+### Why the fallacy?
+
+So _why_ do some people tend to think that it's more efficient to merge new
+code and do damage control later (a.k.a shoot-from-the-hip coding)? Well, in
+the short term, it may save time for the implementer (let's call a spade a
+space: the culprit) by:
+
+* Not writing tests or coming up with enough test cases.
+* Potentially merging without review.
+
+These may save time to one person in the short term by allowing them to merge
+faster and move on to other things; in the greater scheme of things however
+time and energy is wasted.
+
+Another reason why people tend to believe that fixing a bug in master isn't
+necessarily worse than catching it before it is merged is that people often
+don't realize that there's a cost associated with reporting bugs. And this cost
+grows linearily with the number of people subscribed to the bug tracker,
+involved in triage, and QA.
+
+If you like your team, please make sure you catch your bugs before others find
+them themselves!
+
+## Appendix: merge queues for greater good
+
+Bugs are not the only way to break master, race conditions also apply to
+software process. They may happen when a CI system only runs tests on a branch,
+without rebasing on master first. The following diagram lists three branches,
+`master`, `bob` and `alice`:
```

Comment:
AFAIK know some tools (Travis) runs two tests for every PR:

- `push` (the code that was pushed, i.e. the HEAD of the branch that you want to merge)
- `pr` (the code that you would get by merging this branch into master)

similar terminology for the same idea, I think it's worth mentioning that common tools may already do this